### ISSUE TITLE
support async functions in `zig test`

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#3117]
```zig
pub fn foo() void {
    suspend;
}

test "foo" {
    foo();
}
```

```
1/1 test "foo"...resumed an async function which already returned
/Users/benjamin.feng/projects/zig-toy/async.zig:5:1: 0x10bd8257d in _test "foo" (test.o)
test "foo" {
^
/usr/local/Cellar/zig/0.4.0+async/lib/zig/std/special/test_runner.zig:13:25: 0x10bda4155 in _std.special.main (test.o)
        if (test_fn.func()) |_| {
                        ^
/usr/local/Cellar/zig/0.4.0+async/lib/zig/std/special/start.zig:139:37: 0x10bda3f14 in _main (test.o)
            const result = root.main() catch |err| {
```

What's happening here is that Zig is putting function pointers into a list and invoking undefined behavior because the test runner is calling all the test functions as if they were normal functions.

A first step would be adding a compile error for these functions to stop this UB. (bug)

The next step is making it so the test runner knows whether a test function is async or not, and it can do what it needs to do to run async tests. (enhancement)

**Comment by andrewrk** [comment]
After 77a5f888be664f9ef09e2c93f52338448e992e00

```
/home/andy/dev/zig/build/test.zig:5:1: error: test functions cannot be async
test "foo" {
^
/home/andy/dev/zig/build/test.zig:5:1: note: this restriction may be lifted in the future. See https://github.com/ziglang/zig/issues/3117 for more details
test "foo" {
^
/home/andy/dev/zig/build/test.zig:6:8: note: async function call here
    foo();
       ^
/home/andy/dev/zig/build/test.zig:2:5: note: suspends here
    suspend;
    ^
```

**Comment by andrewrk** [comment]
Solved with the merge of #4404, landed in 014f66e6de4aaf81f32c796b12f981326a479397.
